### PR TITLE
Fix lock modes

### PR DESCRIPTION
**Pull request by adamretter**
Fixes a bunch of incorrect Lock mode issues.

### CODE REVIEW

File: extensions/webdav/src/org/exist/webdav/ExistCollection.java [hunk 1]
```diff
@@ -443,6 +443,7 @@ void resourceCopyMove(XmldbURI destCollectionUri, String newName, Mode mode) thr
         }
 
         Collection srcCollection = null;
+        final LockMode srcCollectionMode = (mode == Mode.COPY ? LockMode.READ_LOCK : LockMode.WRITE_LOCK);
```
Comment (shabanovd) [inline extensions/webdav/src/org/exist/webdav/ExistCollection.java]: variable name unclear. `srcCollectionLockMode`?

File: extensions/webdav/src/org/exist/webdav/ExistDocument.java [hunk 2]
```diff
@@ -531,6 +531,7 @@ void resourceCopyMove(XmldbURI destCollectionUri, String newName, Mode mode) thr
         }
 
         Collection srcCollection = null;
+        final LockMode srcCollectionMode = (mode == Mode.COPY ? LockMode.READ_LOCK : LockMode.WRITE_LOCK);
```
Comment (shabanovd) [inline extensions/webdav/src/org/exist/webdav/ExistDocument.java]: `srcCollectionLockMode`?

File: src/org/exist/protocolhandler/embedded/EmbeddedUpload.java [hunk 3]
```diff
@@ -102,7 +103,7 @@ public void stream(XmldbURL xmldbURL, InputStream is, Subject user) throws IOExc
      * @param tmp Document that is inserted.
      * @throws IOException
      */
-    public void stream(XmldbURL xmldbURL, File tmp) throws IOException {
+    public void stream(XmldbURL xmldbURL, Path tmp) throws IOException {
```
Comment (shabanovd) [inline src/org/exist/protocolhandler/embedded/EmbeddedUpload.java]: That is API change!

File: src/org/exist/protocolhandler/embedded/EmbeddedUpload.java [hunk 4]
```diff
@@ -115,17 +116,11 @@ public void stream(XmldbURL xmldbURL, File tmp) throws IOException {
      * @param tmp Document that is inserted.
      * @throws IOException
      */
-    public void stream(XmldbURL xmldbURL, File tmp, Subject user) throws IOException {
+    public void stream(XmldbURL xmldbURL, Path tmp, Subject user) throws IOException {
```
Comment (shabanovd) [inline src/org/exist/protocolhandler/embedded/EmbeddedUpload.java]: That is API change!

File: src/org/exist/protocolhandler/embedded/EmbeddedUpload.java [hunk 5]
```diff
@@ -167,41 +163,36 @@ public void stream(XmldbURL xmldbURL, File tmp, Subject user) throws IOException
 
                     if (mime.isXMLType()) {
                         LOG.debug("storing XML resource");
-                        final InputSource inputsource = new InputSource(tmp.toURI().toASCIIString());
+                        final InputSource inputsource = new InputSource(tmp.toUri().toASCIIString());
```
Comment (shabanovd) [inline src/org/exist/protocolhandler/embedded/EmbeddedUpload.java]: use `org.exist.util.FileInputSource`?

File: src/org/exist/protocolhandler/embedded/EmbeddedUpload.java [hunk 6]
```diff
@@ -167,41 +163,36 @@ public void stream(XmldbURL xmldbURL, File tmp, Subject user) throws IOException
...
-                        collectionLocked = false;
                         collection.store(txn, broker, info, inputsource);
                         LOG.debug("done");
 
                     } else {
                         LOG.debug("storing Binary resource");
-                        try (final InputStream is = new FileInputStream(tmp)) {
-                            collection.addBinaryResource(txn, broker, documentUri, is, contentType, tmp.length());
+                        try (final InputStream is = Files.newInputStream(tmp)) {
+                            collection.addBinaryResource(txn, broker, documentUri, is, contentType, FileUtils.sizeQuietly(tmp));
```
Comment (shabanovd) [inline src/org/exist/protocolhandler/embedded/EmbeddedUpload.java]: how `FileUtils.sizeQuietly` good here?

File: extensions/webdav/src/org/exist/webdav/ExistCollection.java (hunk 1 above)
Comment (adamretter) [inline extensions/webdav/src/org/exist/webdav/ExistCollection.java]: Fixed.

File: extensions/webdav/src/org/exist/webdav/ExistDocument.java (hunk 2 above)
Comment (adamretter) [inline extensions/webdav/src/org/exist/webdav/ExistDocument.java]: Fixed.

File: src/org/exist/protocolhandler/embedded/EmbeddedUpload.java (hunk 3 above)
Comment (adamretter) [inline src/org/exist/protocolhandler/embedded/EmbeddedUpload.java]: Ah right, sorry I cherry-picked a commit which should not have been included. I have now removed it.

Comment (adamretter) [inline src/org/exist/protocolhandler/embedded/EmbeddedUpload.java]: Ah right, sorry I cherry-picked a commit which should not have been included. I have now removed it.

File: src/org/exist/protocolhandler/embedded/EmbeddedUpload.java (hunk 4 above)
Comment (adamretter) [inline src/org/exist/protocolhandler/embedded/EmbeddedUpload.java]: Ah right, sorry I cherry-picked a commit which should not have been included. I have now removed it.

File: src/org/exist/protocolhandler/embedded/EmbeddedUpload.java (hunk 5 above)
Comment (adamretter) [inline src/org/exist/protocolhandler/embedded/EmbeddedUpload.java]: Dropped the change

File: src/org/exist/protocolhandler/embedded/EmbeddedUpload.java (hunk 6 above)
Comment (adamretter) [inline src/org/exist/protocolhandler/embedded/EmbeddedUpload.java]: Dropped the change

**Comment by joewiz** [comment]
@dizzzz Do you think this might help WebDAV and/or replication by chance?

**Comment by dizzzz** [comment]
@joewiz WebDAV for sure

**Comment by adamretter** [comment]
@joewiz it will probably help everything; One of the issues was that `NativeBroker#getResource` would have been taking an Exclusive lock on a Collection when opening a document for writing, however it only needed a Shared Lock.

**Comment by dizzzz** [comment]
@adamretter I realise you made some significant changes in the webdav code regarding LOCKs. In the replication extension there are similar operations. I'll review[ this code myself](https://github.com/dizzzz/messaging-replication/blob/develop/java/src/org/exist/jms/replication/subscribe/ReplicationJmsListener.java) but please could you help me te review the code for similar issues?

**Comment by adamretter** [comment]
@dizzzz I would love to help, but right now that is kinda near the bottom of a very long TODO list for me. Maybe give me a poke in a few weeks about it?

**Comment by dizzzz** [comment]
@adamretter yes sure :) I think there are not too many issues there, but I saw some improvements already...

**Comment by adamretter** [comment]
@dizzzz @shabanovd so can we merge this then?

**Pull request merged**